### ISSUE TITLE
request help: Accessing ctx.var.upstream_uri in global rule causes failure of proxy-rewrite

### ISSUE THREAD

**Issue by Gary-Airwallex** [apache/apisix#3903]
### Issue description
Minimum steps to reproduce:
1. Add a plugin that contains this line in rewrite phase:
```
core.log.info("Access upstream uri: ", ctx.var.upstream_uri)
```
2. Add the plugin to global rule.
3. Add a route like this:
```
"uri": "/hello",
"plugins": {
    "proxy-rewrite": {
        "uri": "/world"
    }
}
```
4. Hit `/hello`, observe that it won't be rewritten to `/world`

Essentially after ctx.var.upstream_uri is set in proxy-rewrite, it fails to be updated in ngx.var, given that we read the value ctx.var.upstream_uri in global rule.
The case was found in APISIX 2.0 so it might not apply to the current master version (I can see that init.lua has changed a lot). If someone know the root cause of this I'd be grateful to be enlightened. 

### Environment

* apisix version (cmd: `apisix version`): 2.0
* OS (cmd: `uname -a`): osx & ubuntu
* OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`): 1.15.8
* etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API): 3.4.9
* apisix-dashboard version, if have: N/A

**Comment by Firstsawyou** [comment]
Not sure if this problem is fixed, maybe you can try the latest version.
Let me check if I have this problem in the latest version.

**Comment by Firstsawyou** [comment]
Hi, @Gary-Airwallex 
I did not reproduce this problem locally. I will provide a recurring use case later.

**Comment by Firstsawyou** [comment]
Hi, @Gary-Airwallex  Sorry, my previous test method was incorrect. I checked the latest version again, and there seems to be such a problem.   @spacewander  Please take a look at this question when you are free.

Here are the steps to reproduce:

1. Add the log of `ctx.var.upstream_uri` in the rewrite phase of the `cors` plugin, and print the rewritten uri in the `proxy-rewrite` plugin.

```shell
diff --git a/apisix/plugins/cors.lua b/apisix/plugins/cors.lua
index 4b0b7c4..98beecc 100644
--- a/apisix/plugins/cors.lua
+++ b/apisix/plugins/cors.lua
@@ -225,6 +225,7 @@ end
 
 
 function _M.rewrite(conf, ctx)
+    core.log.warn("Access upstream uri: ", ctx.var.upstream_uri)
     if ctx.var.request_method == "OPTIONS" then
         return 200
     end
diff --git a/apisix/plugins/proxy-rewrite.lua b/apisix/plugins/proxy-rewrite.lua
index c678b59..e0c3d58 100644
--- a/apisix/plugins/proxy-rewrite.lua
+++ b/apisix/plugins/proxy-rewrite.lua
@@ -179,6 +179,7 @@ function _M.rewrite(conf, ctx)
     end
 
     if not conf.headers then
+        core.log.warn("upstream_uri: ", ctx.var.upstream_uri)
         return
     end
```

2. Add the global plugin `cors`

```shell
curl -X PUT http://127.0.0.1:9080/apisix/admin/global_rules/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -d '{
        "plugins": {
           "cors": {}
        }
}'
```

3. Add a route and bind the `proxy-rewrite` plugin

```shell
curl http://127.0.0.1:9080/apisix/admin/routes/1 -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "methods": ["GET"],
    "uri": "/hello",
    "plugins": {
        "proxy-rewrite": {
            "uri": "/route01"
        }
    },
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "127.0.0.1:1990": 1
        }
    }
}'
```

4. Test route (upstream default response data is the requested uri)

```shell
curl http://127.0.0.1:9080/hello -i
HTTP/1.1 200 OK
Content-Type: application/octet-stream
Transfer-Encoding: chunked
Connection: keep-alive
Date: Thu, 25 Mar 2021 16:05:58 GMT
Server: APISIX/2.4
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: *
Access-Control-Max-Age: 5
Access-Control-Expose-Headers: *
Access-Control-Allow-Headers: *

request_uri: /hello
```

The `proxy-rewrite` plug-in rewriting uri is not effective.

`access.log` information:

```
[warn] 2707#2707: *62368 [lua] proxy-rewrite.lua:182: phase_func(): upstream_uri: /route01, client: 127.0.0.1,
```

5. Remove the log of `ctx.var.upstream_uri` in the global plugin `cors`

```shell
diff --git a/apisix/plugins/proxy-rewrite.lua b/apisix/plugins/proxy-rewrite.lua
index c678b59..e0c3d58 100644
--- a/apisix/plugins/proxy-rewrite.lua
+++ b/apisix/plugins/proxy-rewrite.lua
@@ -179,6 +179,7 @@ function _M.rewrite(conf, ctx)
     end
 
     if not conf.headers then
+        core.log.warn("upstream_uri: ", ctx.var.upstream_uri)
         return
     end
```

6. Test the route again (the upstream default response data is the requested uri)

```shell
$ curl http://127.0.0.1:9080/hello -i
HTTP/1.1 200 OK
Content-Type: application/octet-stream
Transfer-Encoding: chunked
Connection: keep-alive
Date: Thu, 25 Mar 2021 16:13:31 GMT
Server: APISIX/2.4
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: *
Access-Control-Max-Age: 5
Access-Control-Expose-Headers: *
Access-Control-Allow-Headers: *

request_uri: /route01
```
The `proxy-rewrite` plug-in rewrites the uri to take effect.

`access.log` information:

```
[warn] 2707#2707: *62368 [lua] proxy-rewrite.lua:182: phase_func(): upstream_uri: /route01, client: 127.0.0.1,
```

**Comment by spacewander** [comment]
It is because `ctx.var` caches the result:
https://github.com/apache/apisix/blob/83931ba0a5c95e04ec867a4ffd0821d426ca0d27/apisix/core/ctx.lua#L186

**Comment by spacewander** [comment]
One solution is to use ngx.var instead:
https://github.com/apache/apisix/pull/3914

**Comment by spacewander** [comment]
Another solution is to use a new way to cache `ctx.var`:
https://github.com/apache/apisix/pull/3915

This way has performance slow down, CC @membphis

**Comment by Firstsawyou** [comment]
> It is because `ctx.var` caches the result:
> https://github.com/apache/apisix/blob/83931ba0a5c95e04ec867a4ffd0821d426ca0d27/apisix/core/ctx.lua#L186

Got it, thank you.

**Comment by zwjzxh520** [comment]
臣附议

**Comment by zwjzxh520** [comment]
@Firstsawyou 你好，请问这个问题怎么修复呢？

**Comment by Firstsawyou** [comment]
Hi, @zwjzxh520 
@spacewander  provides two solutions, you can refer to the following:

Method one: use `ngx.var`  instead `ctx.var`, for example: `ngx.var.upstream_uri` .  For detailed examples, you can refer to #3914 

Method two:  use a new way to cache ctx.var: #3915

**Comment by zwjzxh520** [comment]
懂了，难道我一直在 #3914 里找不到修复的源码，只看到增加的测试用例。谢谢

**Comment by spacewander** [comment]
Solved by #3915
